hyper-capi/examples/client.c

```diff
@@ -0,0 +1,303 @@
+			return HYPER_IO_PENDING;
+		} else {
+			// kaboom
+			return HYPER_IO_ERROR;
+		}
+	} else {
+		return ret;
+	}
+}
+
+static int connect_to(const char *host, const char *port) {
+	struct addrinfo hints;
+	memset(&hints, 0, sizeof(struct addrinfo));
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+
+	struct addrinfo *result, *rp;
+	if (getaddrinfo(host, port, &hints, &result) != 0) {
+		printf("dns failed for %s\n", host);
+		return -1;
+	}
+
+	int sfd;
+	for (rp = result; rp != NULL; rp = rp->ai_next) {
+		sfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+		if (sfd == -1) {
+			continue;
+		}
+
+		if (connect(sfd, rp->ai_addr, rp->ai_addrlen) != -1) {
+			break;
+		} else {
+			close(sfd);
+		}
+	}
+
+	freeaddrinfo(result);
+
+	// no address succeeded
+	if (rp == NULL) {
+		printf("connect failed for %s\n", host);
+		return -1;
+	}
+
+	return sfd;
+}
+
+static hyper_iter_step print_each_header(void *userdata,
+                                         const uint8_t *name,
+                                         size_t name_len,
+                                         const uint8_t *value,
+                                         size_t value_len) {
+	printf("%.*s: %.*s\n", (int) name_len, name, (int) value_len, value);
+	return HYPER_ITER_CONTINUE;
+}
+
+static hyper_iter_step print_each_chunk(void *userdata, const hyper_buf *chunk) {
+	const uint8_t *buf = hyper_buf_bytes(chunk);
+	size_t len = hyper_buf_len(chunk);
+
+	write(1, buf, len);
+
+	return HYPER_ITER_CONTINUE;
+}
+
+typedef enum {
+	EXAMPLE_NOT_SET = 0, // tasks we don't know about won't have a userdata set
+	EXAMPLE_HANDSHAKE,
+	EXAMPLE_SEND,
+	EXAMPLE_RESP_BODY
+} example_id;
+
+#define STR_ARG(XX) (uint8_t *)XX, strlen(XX)
+
+int main(int argc, char *argv[]) {
+        const char *host = argc > 1 ? argv[1] : "httpbin.org";
+        const char *port = argc > 2 ? argv[2] : "80";
+        const char *path = argc > 3 ? argv[3] : "/";
+        printf("connecting to port %s on %s...\n", port, host);
+
+        int fd = connect_to(host, port);
+	if (fd < 0) {
+		return 1;
+	}
+        printf("connected to %s, now get %s\n", host, path);
+
+	if (fcntl(fd, F_SETFL, O_NONBLOCK) != 0) {
+		printf("failed to set socket to non-blocking\n");
+		return 1;
+	}
+
+	fd_set fds_read;
+	fd_set fds_write;
+	fd_set fds_excep;
+
+	struct conn_data *conn = malloc(sizeof(struct conn_data));
+
+	conn->fd = fd;
+	conn->read_waker = NULL;
+	conn->write_waker = NULL;
+
+
+	// Hookup the IO
+	hyper_io *io = hyper_io_new();
+	hyper_io_set_data(io, (void *)conn);
+	hyper_io_set_read(io, read_cb);
+	hyper_io_set_write(io, write_cb);
+
+	printf("http handshake ...\n");
+
+	// We need an executor generally to poll futures
+	hyper_executor *exec = hyper_executor_new();
+
+	// Prepare client options
+	hyper_clientconn_options *opts = hyper_clientconn_options_new();
+	hyper_clientconn_options_exec(opts, exec);
+
+	hyper_task *handshake = hyper_clientconn_handshake(io, opts);
```

Comment:
`handshake` is never freed later in the example code. Is that a mistake or does hyper handle that itself somehow?

Reply:
Ah no, now I think I get it. That's the `task` returned from `hyper_executor_poll` in the first invoke, right?

Reply:
Correct, `hyper_executor_push` takes ownership of the task. If it gets popped later with `hyper_executor_poll`, that gives ownership back and you can free it. If you were to free the executor while it still owned some tasks, it'd free them at the same time.